python/cudf/udf_cpp/shim.cu

```diff
@@ -662,3 +664,41 @@ make_definition_idx(BlockIdxMax, int64, int64_t);
 make_definition_idx(BlockIdxMax, float64, double);
 #undef make_definition_idx
 }
+
+__device__ void udf_str_dtor(void* udf_str, size_t size, void* dtor_info)
+{
+  udf_string* ptr = reinterpret_cast<udf_string*>(udf_str);
+  ptr->~udf_string();
+}
+
+// Only used to allocate the right amount of space, see below
+struct meminfo_and_str {
+  NRT_MemInfo mi;
+  udf_string st;
+};
+
+/*
+Create a new MemInfo object holding the reference count of a udf_string. When returning
+new strings, shim functions expect a pointer to a stack allocated buffer into which it
```

Comment:
> shim functions expect a pointer to a stack allocated buffer

Is this right? Or is it that we happen to always pass pointers to stack allocated objects to the shim function?

Reply:
What buffers are we referring to here? Everything in the `udf_string` is heap allocated AFAIR.

Reply:
This refers to the piece of memory containing the `udf_string` struct itself. These are allocated on the numba side before being passed in, for example [here](https://github.com/rapidsai/cudf/blob/branch-23.06/python/cudf/cudf/core/udf/strings_lowering.py#L228). More info on `alloca` [here](https://llvm.org/docs/LangRef.html#alloca-instruction). The important bit being

```
The ‘alloca’ instruction allocates sizeof(<type>)*NumElements bytes of memory on the runtime stack, returning a pointer of the appropriate type to the program
```

The `udf_string` consists of three members, all _technically_ integers and is usually 24 bytes IIRC:
- the length of the string
- the total number of bytes in the string
- a pointer to the heap allocation containing the string

So while one of the struct members is a pointer to a heap allocation the struct itself lives in a stack allocation. This is the reason so much fancy footwork is needed in this function. The `NRT_MemInfo` object must be created around a pointer of some kind. It can't be created around the pointer to the stack allocation containing the `udf_string`, because later on when it comes time to call the destructor it may not be there anymore.

One could use the heap pointer underneath the `udf_string` for this purpose technically, since the pointed-to allocation won't be going anywhere until explicitly freed. But in discussing this with @davidwendt we decided that it's not a great idea to be freeing pointers underneath a `udf_string` via some external mechanism that isn't exactly the `udf_string` dtor itself. For instance imagine in the future if `udf_string` was changed somehow to require the other members to correctly clean up all of its data. At that point just having a pointer but no way of getting to the rest of the `udf_string` won't be sufficient.

Reply:
(note - I hadn't got Brandon's previous comment visible when I posted mine, so ignore mine for now)

Reply:
Agreed, we don't want to reach into the `udf_string` and rely on its internals. The design here makes sense now that you've explained it, I needed the pointers to the builder and the realization that we're talking about stack allocations within the context of the LLVM instructions rather than in the code actually written here. The explanation is super helpful, thanks.